### ISSUE TITLE
[a11y][Mobile][iOS][WCAG 2.4.8 ] [v2.1 AA] Focus indicator remains persistent on some components

### ISSUE THREAD

**Issue by dkatzz** [civiform/civiform#9911]
OBSERVED RESULTS:
--

While using VoiceOver on iOS, there are certain components where the focus indicator persist even when the element is not currently being focused by the user, as shown in the following video where linear navigation is conducted and some components display incorrectly the focus indicator

[iOS](https://drive.google.com/file/d/1bjGTVI9FoAudxxSew2cZPLYnK8iAC_5x/view?usp=drive_link)

EXPECTED RESULTS:
--

Information about the user's location within the page should be accurate as to avoid confusion about the current location of selection of components.

STEPS TO REPRODUCE:
--
0. Turn on the screen reader (VoiceOver)
1. Start the navigation on a linear manner 
2. Pay attention to the focus indicator, notice how there are certain elements that display a persistent focus indicator.
3. Focus indicator only shifts to the next component once a new component with the capacity of displaying the focus indicator is selected. 

RELEVANT GAR SHEET:
--

[GAR sheet](https://docs.google.com/spreadsheets/d/1mgAhVZ3tuJ0O2EeB5JZ0HAvSrxVeIUjlJxVAhu3JH7M/edit?gid=1240238308#gid=1240238308)

ADDITIONAL INFORMATION:
--

[Success Criterion 2.4.8 Location](https://www.w3.org/TR/WCAG22/#focus-visible)

ENVIRONMENT DETAILS:
--

---
- Devices: iPhone 14 plus: iOS 18.2.1
- Screen reader name: VoiceOver
- Browser name: Chrome Mobile: v133.0.69
---

**Comment by melanie-exygy** [comment]
I've been testing this on the W3C website and on the Deque University website.  I tested with VoiceOver on my Desktop.  In both cases, the focus indicator stays on the last interactive item while linear navigation moves forward.  Focus can only be sent to focusable items.  I believe this is expected behavior.

https://github.com/user-attachments/assets/223c39c5-2f39-4574-8693-7914f8feca63

That said, in the iPhone test video in the bug description, focus doesn't shift to the "End session" link or the "View and apply" buttons, which are interactive items.  I think this is the real bug, rather than "Focus indicator only shifts to the next component once a new component with the capacity of displaying the focus indicator is selected," which is normal behavior.

**Comment by melanie-exygy** [comment]
> For some references, W3C has the same behavior as Civiform, and Deque University doesn't have any instances of focus in voiceover mode in iOS.
> 
> Deque University: https://github.com/user-attachments/assets/015ea8a6-c825-4589-90f0-62707d11a2bc W3C: https://github.com/user-attachments/assets/a066d3ea-f692-4ce9-bcb8-1dce3f9d0e40

Interesting that on the W3C site, the focus also stays on "Menu" and doesn't go to the links.  That really is just like CiviForm's behavior.  

For the Deque one, I did notice the blue focus indicator around the last three large buttons ("Full access", "Web accessibility" and "IAAP Accessibility Certification"), but it's definitely interesting that the focus doesn't appear before that.

**Comment by tallulahkay** [comment]
> For the Deque one, I did notice the blue focus indicator around the last three large buttons ("Full access", "Web accessibility" and "IAAP Accessibility Certification"), but it's definitely interesting that the focus doesn't appear before that.

Didn't notice this! So true!

**Comment by dkatzz** [comment]
@melanie-exygy here is the reply from the a11y reviewer (note they did desktop testing for the most part and then followed up with some mobile testing):

Since this was supposed to be a mobile testing I preformed using only the tools inside the mobile device that means: swipe right, swipe left, double-tap all with the screen reader on.

I believe I see what you mean, so the focus can that is independent from the screen reader cannot be shifted to components that cannot be interacted with and that's why it remains persistent on some elements, and I believe I did mention that if you intention was to keep the focus on every single element even if some of them were just text or headings then there may be an issue with the focus however take this example:

if you check this [wikipedia](https://en.wikipedia.org/wiki/Computer_accessibility#bodyContent) page you can jump from element to element using the "Tab" key and it should only focus elements that can be interacted with, however in the page that I tested that was not the case as every single element was focusable, using the screen reader there is a way to do that but is using "control + option + arrow keys" I did not want to file a bug for it since maybe it was intended to work that way, however if that's the case then we should try to fix the elements that do display a highlight focus that remains in place if they are not currently being selected

[Example page](https://screencast.googleplex.com/cast/NDkwMjc2MzQ5MzMyNjg0OHw1ODdjNDAyNi1mZA)

And also yes there are certain elements that are supposed to receive the highlight focus because they are intractable elements and they are not getting it, I did not file a bug for them since the fix for this issue with the focus should also address those components so it would be redundant or "duplicated"

If you have further questions we could set up a call or just ask again here

**Comment by melanie-exygy** [comment]
After further testing on iOS, @tallulahkay and I have discovered that keyboard focus doesn't follow the VoiceOver cursor to links.  However, it does follow on buttons.  This is consistent behavior with what we observed on the Deque University and W3C websites.  So, any link with a `role="button"` will receive keyboard focus when the cursor moves to that element.  This is why the "Dev Tools" and "Select language" buttons receive focus.  However, any link with a link role will not receive focus.

It is important that any elements that truly serve the purpose of a link, which is to allow the user to navigate to a different section of the page or to a different page, retain the "link" aria-role.  This is because screen reader users will often navigate by links or get a list of the links on the page and if we give those anchor tags the role of button, screen readers will no longer include those links in the link navigation menu. 

This means that links like "View and apply" need to remain links in spite of them not receiving focus in iOS.  Note that iOS users will still be able to navigate to links using the VoiceOver cursor and will still be able to click on those links by double-tapping.  Also note that focus does appear around the link when the user double-taps.
